File: packages/lynx/android/src/main/java/dev/taurinative/lynx/TauriNativeRust.java
package dev.taurinative.lynx;

import java.nio.charset.StandardCharsets;

final class TauriNativeRust {
  static {
    System.loadLibrary("taurinativelynx");
  }

  private TauriNativeRust() {}

  static String invoke(String command, String payloadJson) {
    byte[] response = invokeBytes(
      command.getBytes(StandardCharsets.UTF_8),
      payloadJson.getBytes(StandardCharsets.UTF_8)
    );
    return new String(response, StandardCharsets.UTF_8);
  }

  private static native byte[] invokeBytes(byte[] command, byte[] payloadJson);
  static native long createSession();
  static String start(long session, String id, String command, String payload) {
    return new String(startBytes(session, id.getBytes(StandardCharsets.UTF_8), command.getBytes(StandardCharsets.UTF_8), payload.getBytes(StandardCharsets.UTF_8)), StandardCharsets.UTF_8);
  }
  static String poll(long session) { return new String(pollBytes(session), StandardCharsets.UTF_8); }
  static void cancel(long session, String id) { cancelBytes(session, id.getBytes(StandardCharsets.UTF_8)); }
  static native void closeSession(long session);
  private static native byte[] startBytes(long session, byte[] id, byte[] command, byte[] payload);
  private static native byte[] pollBytes(long session);
  private static native void cancelBytes(long session, byte[] id);
}

File: packages/react-native/android/src/main/java/com/reactnativetauri/TauriNativeRust.kt
package com.reactnativetauri

internal object TauriNativeRust {
  init {
    System.loadLibrary("reactnativetauri")
  }

  fun invoke(command: String, payloadJson: String): String = String(
    invokeBytes(
      command.toByteArray(Charsets.UTF_8),
      payloadJson.toByteArray(Charsets.UTF_8),
    ),
    Charsets.UTF_8,
  )

  private external fun invokeBytes(
    command: ByteArray,
    payloadJson: ByteArray,
  ): ByteArray

  external fun createSession(): Long
  fun start(session: Long, id: String, command: String, payload: String): String = String(
    startBytes(session, id.toByteArray(Charsets.UTF_8), command.toByteArray(Charsets.UTF_8), payload.toByteArray(Charsets.UTF_8)), Charsets.UTF_8,
  )
  fun poll(session: Long): String = String(pollBytes(session), Charsets.UTF_8)
  fun cancel(session: Long, id: String) = cancelBytes(session, id.toByteArray(Charsets.UTF_8))
  external fun closeSession(session: Long)
  private external fun startBytes(session: Long, id: ByteArray, command: ByteArray, payload: ByteArray): ByteArray
  private external fun pollBytes(session: Long): ByteArray
  private external fun cancelBytes(session: Long, id: ByteArray)
}
